File: frontend/src/types/resources.ts
// =============================================================================
// Application Info
// =============================================================================

export interface AppInfo {
  version: string;
  build_sha: string;
  build_date: string;
  environment: string;
  timestamp: string;
}

// =============================================================================
// Resource Types
// =============================================================================

export type DisplayStatus =
  | "active"
  | "inactive"
  | "transitioning"
  | "error"
  | "unknown";

export type ManagedBy = "terraform" | "github_actions" | "unmanaged";

export interface EC2Instance {
  id: number;
  instance_id: string;
  name: string | null;
  instance_type: string;
  state: string;
  display_status: DisplayStatus;
  private_ip: string | null;
  public_ip: string | null;
  private_dns: string | null;
  public_dns: string | null;
  vpc_id: string | null;
  subnet_id: string | null;
  availability_zone: string | null;
  launch_time: string | null;
  tags: Record<string, string> | null;
  tf_managed: boolean;
  tf_state_source: string | null;
  tf_resource_address: string | null;
  region_name: string | null;
  is_deleted: boolean;
  deleted_at: string | null;
  updated_at: string;
  created_at?: string;
}

export interface RDSInstance {
  id: number;
  db_instance_identifier: string;
  name: string | null;
  db_instance_class: string;
  status: string;
  display_status: DisplayStatus;
  engine: string;
  engine_version: string;
  allocated_storage: number;
  endpoint: string | null;
  port: number | null;
  vpc_id: string | null;
  availability_zone: string | null;
  multi_az: boolean;
  tags: Record<string, string> | null;
  tf_managed: boolean;
  tf_state_source: string | null;
  tf_resource_address: string | null;
  region_name: string | null;
  is_deleted: boolean;
  deleted_at: string | null;
  updated_at: string;
  created_at?: string;
}

export interface VPC {
  id: number;
  vpc_id: string;
  name: string | null;
  cidr_block: string;
  state: string;
  is_default: boolean;
  display_status: DisplayStatus;
  enable_dns_support: boolean;
  enable_dns_hostnames: boolean;
  tags: Record<string, string> | null;
  tf_managed: boolean;
  tf_state_source: string | null;
  tf_resource_address: string | null;
  region_name: string | null;
  is_deleted: boolean;
  deleted_at: string | null;
  updated_at: string;
  created_at?: string;
}

export type SubnetType = "public" | "private" | "unknown";

export interface Subnet {
  id: number;
  subnet_id: string;
  name: string | null;
  vpc_id: string;
  cidr_block: string;
  availability_zone: string;
  subnet_type: SubnetType;
  state: string;
  display_status: DisplayStatus;
  available_ip_count: number;
  map_public_ip_on_launch: boolean;
  tags: Record<string, string> | null;
  tf_managed: boolean;
  tf_state_source: string | null;
  tf_resource_address: string | null;
  region_name: string | null;
  is_deleted: boolean;
  deleted_at: string | null;
  updated_at: string;
  created_at?: string;
}

export interface InternetGateway {
  id: number;
  igw_id: string;
  name: string | null;
  vpc_id: string | null;
  state: string;
  display_status: DisplayStatus;
  tags: Record<string, string> | null;
  tf_managed: boolean;
  tf_state_source: string | null;
  tf_resource_address: string | null;
  region_name: string | null;
  is_deleted: boolean;
  deleted_at: string | null;
  updated_at: string;
  created_at?: string;
}

export type ConnectivityType = "public" | "private";

export interface NATGateway {
  id: number;
  nat_gateway_id: string;
  name: string | null;
  vpc_id: string;
  subnet_id: string;
  state: string;
  connectivity_type: ConnectivityType;
  display_status: DisplayStatus;
  primary_private_ip: string | null;
  primary_public_ip: string | null;
  allocation_id: string | null;
  network_interface_id: string | null;
  tags: Record<string, string> | null;
  tf_managed: boolean;
  tf_state_source: string | null;
  tf_resource_address: string | null;
  region_name: string | null;
  is_deleted: boolean;
  deleted_at: string | null;
  updated_at: string;
  created_at?: string;
}

export interface ElasticIP {
  id: number;
  allocation_id: string;
  name: string | null;
  public_ip: string;
  private_ip: string | null;
  domain: string;
  display_status: DisplayStatus;
  association_id: string | null;
  instance_id: string | null;
  network_interface_id: string | null;
  tags: Record<string, string> | null;
  tf_managed: boolean;
  tf_state_source: string | null;
  tf_resource_address: string | null;
  region_name: string | null;
  is_deleted: boolean;
  deleted_at: string | null;
  updated_at: string;
  created_at?: string;
}

export type ECSLaunchType = "FARGATE" | "EC2" | "EXTERNAL";

export interface ECSContainer {
  id: number;
  task_id: string;
  name: string | null;
  cluster_name: string;
  launch_type: ECSLaunchType;
  status: string;
  display_status: DisplayStatus;
  cpu: number;
  memory: number;
  task_definition_arn: string | null;
  desired_status: string | null;
  image: string | null;
  image_tag: string | null;
  container_port: number | null;
  private_ip: string | null;
  subnet_id: string | null;
  vpc_id: string | null;
  availability_zone: string | null;
  started_at: string | null;
  tags: Record<string, string> | null;
  tf_managed: boolean;
  tf_state_source: string | null;
  tf_resource_address: string | null;
  managed_by: ManagedBy;
  region_name: string | null;
  is_deleted: boolean;
  deleted_at: string | null;
  updated_at: string;
  created_at?: string;
}

export interface ECSClusterSummary {
  cluster_name: string;
  total_tasks: number;
  running_tasks: number;
  stopped_tasks: number;
  pending_tasks: number;
  tf_managed: boolean;
  managed_by: ManagedBy;
  region_name: string | null;
  containers: ECSContainer[];
}

export interface ECSSummaryResponse {
  clusters: number;
  services: number;
  running_tasks: number;
  stopped_tasks: number;
  pending_tasks: number;
  total_tasks: number;
}

// =============================================================================
// API Response Types
// =============================================================================

export interface MetaInfo {
  total: number;
  last_refreshed: string | null;
}

export interface ListResponse<T> {
  data: T[];
  meta: MetaInfo;
}

export interface ResourceCount {
  active: number;
  inactive: number;
  transitioning: number;
  error: number;
  total: number;
}

export interface StatusSummary {
  ec2: ResourceCount;
  rds: ResourceCount;
  total: ResourceCount;
  last_refreshed: string | null;
}

export interface RefreshResponse {
  success: boolean;
  message: string;
  resources_updated: number;
  duration_seconds: number;
}

// =============================================================================
// Terraform Types
// =============================================================================

export interface TerraformBucketInfo {
  name: string;
  region: string | null;
}

export interface TerraformStateInfo {
  name: string;
  key: string;
  bucket: TerraformBucketInfo | null;
  description: string | null;
  last_modified: string | null;
  resource_count: number;
  status: string;
  all_resource_types?: Record<string, number> | null;
  skipped_resource_count?: number;
}

export interface TerraformStatesResponse {
  states: TerraformStateInfo[];
  total_tf_managed_resources: number;
}

export interface DriftItem {
  resource_type: string;
  resource_id: string;
  drift_type: "unmanaged" | "orphaned" | "modified";
  details: string | null;
}

export interface DriftResponse {
  drift_detected: boolean;
  items: DriftItem[];
  checked_at: string;
}

// =============================================================================
// Filter Types
// =============================================================================

export interface ResourceFilters {
  status?: DisplayStatus;
  region?: string;
  search?: string;
  tf_managed?: boolean;
  vpc_id?: string;
  subnet_id?: string;
  subnet_type?: SubnetType;
  connectivity_type?: ConnectivityType;
  instance_id?: string;
  associated?: boolean;
  cluster_name?: string;
  launch_type?: ECSLaunchType;
}

// =============================================================================
// VPC Page Types
// =============================================================================

export type VPCResourceType =
  | "vpcs"
  | "subnets"
  | "internet-gateways"
  | "nat-gateways"
  | "elastic-ips";

File: backend/app/schemas/resources.py
"""
Pydantic schemas for API request/response validation.

These schemas define the shape of data exchanged through the API.
"""

from datetime import datetime
from enum import Enum
from typing import Any, Dict, Generic, List, Optional, TypeVar

from pydantic import BaseModel, ConfigDict, Field


# =============================================================================
# Enums
# =============================================================================


class DisplayStatus(str, Enum):
    """Normalized display status for all resources."""

    ACTIVE = "active"
    INACTIVE = "inactive"
    TRANSITIONING = "transitioning"
    ERROR = "error"
    UNKNOWN = "unknown"


class EC2State(str, Enum):
    """EC2 instance states."""

    RUNNING = "running"
    STOPPED = "stopped"
    PENDING = "pending"
    STOPPING = "stopping"
    SHUTTING_DOWN = "shutting-down"
    TERMINATED = "terminated"


class RDSStatus(str, Enum):
    """RDS instance statuses."""

    AVAILABLE = "available"
    STOPPED = "stopped"
    STARTING = "starting"
    STOPPING = "stopping"
    CREATING = "creating"
    DELETING = "deleting"
    FAILED = "failed"


# =============================================================================
# Base Schemas
# =============================================================================


class BaseSchema(BaseModel):
    """Base schema with common configuration."""

    model_config = ConfigDict(from_attributes=True)


# =============================================================================
# EC2 Schemas
# =============================================================================


class EC2InstanceBase(BaseSchema):
    """Base EC2 instance schema."""

    instance_id: str
    name: Optional[str] = None
    instance_type: str
    state: str
    private_ip: Optional[str] = None
    public_ip: Optional[str] = None
    vpc_id: Optional[str] = None
    subnet_id: Optional[str] = None
    availability_zone: Optional[str] = None


class EC2InstanceResponse(EC2InstanceBase):
    """EC2 instance response schema."""

    id: int
    display_status: DisplayStatus
    launch_time: Optional[datetime] = None
    tags: Optional[Dict[str, Any]] = None
    tf_managed: bool = False
    tf_state_source: Optional[str] = None
    tf_resource_address: Optional[str] = None
    region_name: Optional[str] = Field(None, description="AWS region name")
    is_deleted: bool = False
    deleted_at: Optional[datetime] = None
    updated_at: datetime


class EC2InstanceDetail(EC2InstanceResponse):
    """Detailed EC2 instance response with all fields."""

    created_at: datetime


# =============================================================================
# RDS Schemas
# =============================================================================


class RDSInstanceBase(BaseSchema):
    """Base RDS instance schema."""

    db_instance_identifier: str
    name: Optional[str] = None
    db_instance_class: str
    status: str
    engine: str
    engine_version: str
    allocated_storage: int


class RDSInstanceResponse(RDSInstanceBase):
    """RDS instance response schema."""

    id: int
    display_status: DisplayStatus
    endpoint: Optional[str] = None
    port: Optional[int] = None
    vpc_id: Optional[str] = None
    availability_zone: Optional[str] = None
    multi_az: bool = False
    tags: Optional[Dict[str, Any]] = None
    tf_managed: bool = False
    tf_state_source: Optional[str] = None
    tf_resource_address: Optional[str] = None
    region_name: Optional[str] = Field(None, description="AWS region name")
    is_deleted: bool = False
    deleted_at: Optional[datetime] = None
    updated_at: datetime


class RDSInstanceDetail(RDSInstanceResponse):
    """Detailed RDS instance response with all fields."""

    created_at: datetime


# =============================================================================
# VPC Schemas
# =============================================================================


class VPCBase(BaseSchema):
    """Base VPC schema."""

    vpc_id: str
    name: Optional[str] = None
    cidr_block: str
    state: str
    is_default: bool = False


class VPCResponse(VPCBase):
    """VPC response schema."""

    id: int
    display_status: DisplayStatus
    enable_dns_support: bool = True
    enable_dns_hostnames: bool = False
    tags: Optional[Dict[str, Any]] = None
    tf_managed: bool = False
    tf_state_source: Optional[str] = None
    tf_resource_address: Optional[str] = None
    region_name: Optional[str] = Field(None, description="AWS region name")
    is_deleted: bool = False
    deleted_at: Optional[datetime] = None
    updated_at: datetime


class VPCDetail(VPCResponse):
    """Detailed VPC response with all fields."""

    created_at: datetime


# =============================================================================
# Subnet Schemas
# =============================================================================


class SubnetBase(BaseSchema):
    """Base Subnet schema."""

    subnet_id: str
    name: Optional[str] = None
    vpc_id: str
    cidr_block: str
    availability_zone: str
    subnet_type: str  # public, private, unknown
    state: str


class SubnetResponse(SubnetBase):
    """Subnet response schema."""

    id: int
    display_status: DisplayStatus
    available_ip_count: int = 0
    map_public_ip_on_launch: bool = False
    tags: Optional[Dict[str, Any]] = None
    tf_managed: bool = False
    tf_state_source: Optional[str] = None
    tf_resource_address: Optional[str] = None
    region_name: Optional[str] = Field(None, description="AWS region name")
    is_deleted: bool = False
    deleted_at: Optional[datetime] = None
    updated_at: datetime


class SubnetDetail(SubnetResponse):
    """Detailed Subnet response with all fields."""

    created_at: datetime


# =============================================================================
# Internet Gateway Schemas
# =============================================================================


class InternetGatewayBase(BaseSchema):
    """Base Internet Gateway schema."""

    igw_id: str
    name: Optional[str] = None
    vpc_id: Optional[str] = None  # Null when detached
    state: str


class InternetGatewayResponse(InternetGatewayBase):
    """Internet Gateway response schema."""

    id: int
    display_status: DisplayStatus
    tags: Optional[Dict[str, Any]] = None
    tf_managed: bool = False
    tf_state_source: Optional[str] = None
    tf_resource_address: Optional[str] = None
    region_name: Optional[str] = Field(None, description="AWS region name")
    is_deleted: bool = False
    deleted_at: Optional[datetime] = None
    updated_at: datetime


class InternetGatewayDetail(InternetGatewayResponse):
    """Detailed Internet Gateway response with all fields."""

    created_at: datetime


# =============================================================================
# NAT Gateway Schemas
# =============================================================================


class NATGatewayBase(BaseSchema):
    """Base NAT Gateway schema."""

    nat_gateway_id: str
    name: Optional[str] = None
    vpc_id: str
    subnet_id: str
    state: str
    connectivity_type: str  # public, private


class NATGatewayResponse(NATGatewayBase):
    """NAT Gateway response schema."""

    id: int
    display_status: DisplayStatus
    primary_private_ip: Optional[str] = None
    primary_public_ip: Optional[str] = None
    allocation_id: Optional[str] = None
    network_interface_id: Optional[str] = None
    tags: Optional[Dict[str, Any]] = None
    tf_managed: bool = False
    tf_state_source: Optional[str] = None
    tf_resource_address: Optional[str] = None
    region_name: Optional[str] = Field(None, description="AWS region name")
    is_deleted: bool = False
    deleted_at: Optional[datetime] = None
    updated_at: datetime


class NATGatewayDetail(NATGatewayResponse):
    """Detailed NAT Gateway response with all fields."""

    created_at: datetime


# =============================================================================
# Elastic IP Schemas
# =============================================================================


class ElasticIPBase(BaseSchema):
    """Base Elastic IP schema."""

    allocation_id: str
    name: Optional[str] = None
    public_ip: str
    private_ip: Optional[str] = None
    domain: str  # vpc, standard


class ElasticIPResponse(ElasticIPBase):
    """Elastic IP response schema."""

    id: int
    display_status: DisplayStatus
    association_id: Optional[str] = None
    instance_id: Optional[str] = None
    network_interface_id: Optional[str] = None
    tags: Optional[Dict[str, Any]] = None
    tf_managed: bool = False
    tf_state_source: Optional[str] = None
    tf_resource_address: Optional[str] = None
    region_name: Optional[str] = Field(None, description="AWS region name")
    is_deleted: bool = False
    deleted_at: Optional[datetime] = None
    updated_at: datetime


class ElasticIPDetail(ElasticIPResponse):
    """Detailed Elastic IP response with all fields."""

    created_at: datetime


# =============================================================================
# Terraform Schemas
# =============================================================================


class TerraformStateInfo(BaseSchema):
    """Information about a Terraform state file."""

    name: str
    key: str
    description: Optional[str] = None
    last_modified: Optional[datetime] = None
    resource_count: int = 0
    status: str = "unknown"


class TerraformStatesResponse(BaseSchema):
    """Response for Terraform states endpoint."""

    states: List[TerraformStateInfo]
    total_tf_managed_resources: int


class DriftItem(BaseSchema):
    """A single drift detection item."""

    resource_type: str
    resource_id: str
    drift_type: str  # "unmanaged", "orphaned", "modified"
    details: Optional[str] = None


class DriftResponse(BaseSchema):
    """Response for drift detection endpoint."""

    drift_detected: bool
    items: List[DriftItem]
    checked_at: datetime


# =============================================================================
# Status Summary Schemas
# =============================================================================


class ResourceCount(BaseSchema):
    """Count of resources by status."""

    active: int = 0
    inactive: int = 0
    transitioning: int = 0
    error: int = 0
    total: int = 0


class StatusSummary(BaseSchema):
    """Summary of all resource statuses."""

    ec2: ResourceCount
    rds: ResourceCount
    total: ResourceCount
    last_refreshed: Optional[datetime] = None


# =============================================================================
# Generic Response Schemas
# =============================================================================

T = TypeVar("T")


class PaginatedResponse(BaseSchema, Generic[T]):
    """Generic paginated response wrapper."""

    data: List[T]
    total: int
    page: int = 1
    page_size: int = 50
    has_more: bool = False


class MetaInfo(BaseSchema):
    """Metadata for API responses."""

    total: int
    last_refreshed: Optional[datetime] = None


class ListResponse(BaseSchema, Generic[T]):
    """Generic list response wrapper."""

    data: List[T]
    meta: MetaInfo


# =============================================================================
# Request Schemas
# =============================================================================


class RefreshRequest(BaseSchema):
    """Request to trigger a data refresh."""

    force: bool = Field(
        default=False, description="Force refresh even if recently synced"
    )


class RefreshResponse(BaseSchema):
    """Response from a refresh operation."""

    success: bool
    message: str
    resources_updated: int = 0
    duration_seconds: float = 0.0


# =============================================================================
# Topology Visualization Schemas
# =============================================================================


class TopologyEC2Instance(BaseSchema):
    """EC2 instance for topology visualization."""

    id: str
    name: Optional[str] = None
    instance_type: str
    state: str
    display_status: DisplayStatus
    private_ip: Optional[str] = None
    public_ip: Optional[str] = None
    private_dns: Optional[str] = None
    public_dns: Optional[str] = None
    tf_managed: bool = True
    tf_resource_address: Optional[str] = None


class TopologyRDSInstance(BaseSchema):
    """RDS instance for topology visualization."""

    id: str
    name: Optional[str] = None
    engine: str
    instance_class: str
    status: str
    display_status: DisplayStatus
    endpoint: Optional[str] = None
    port: Optional[int] = None
    tf_managed: bool = True
    tf_resource_address: Optional[str] = None


class TopologyNATGateway(BaseSchema):
    """NAT Gateway for topology visualization."""

    id: str
    name: Optional[str] = None
    state: str
    display_status: DisplayStatus
    primary_public_ip: Optional[str] = None
    tf_managed: bool = True
    tf_resource_address: Optional[str] = None


class TopologyInternetGateway(BaseSchema):
    """Internet Gateway for topology visualization."""

    id: str
    name: Optional[str] = None
    state: str
    display_status: DisplayStatus
    tf_managed: bool = True
    tf_resource_address: Optional[str] = None


class TopologyElasticIP(BaseSchema):
    """Elastic IP for topology visualization."""

    id: str
    public_ip: str
    associated_with: Optional[str] = None
    association_type: Optional[str] = None  # 'ec2', 'nat_gateway', 'eni'
    tf_managed: bool = True
    tf_resource_address: Optional[str] = None


class TopologySubnet(BaseSchema):
    """Subnet with contained resources for topology visualization."""

    id: str
    name: Optional[str] = None
    cidr_block: str
    availability_zone: str
    subnet_type: str  # public, private, unknown
    display_status: DisplayStatus
    tf_managed: bool = True
    tf_resource_address: Optional[str] = None
    nat_gateway: Optional[TopologyNATGateway] = None
    ec2_instances: List[TopologyEC2Instance] = []
    rds_instances: List[TopologyRDSInstance] = []


class TopologyVPC(BaseSchema):
    """VPC with all nested resources for topology visualization."""

    id: str
    name: Optional[str] = None
    cidr_block: str
    state: str
    display_status: DisplayStatus
    tf_managed: bool = True
    tf_resource_address: Optional[str] = None
    internet_gateway: Optional[TopologyInternetGateway] = None
    subnets: List[TopologySubnet] = []
    elastic_ips: List[TopologyElasticIP] = []


class TopologyMeta(BaseSchema):
    """Metadata for topology response."""

    total_vpcs: int = 0
    total_subnets: int = 0
    total_ec2: int = 0
    total_rds: int = 0
    total_nat_gateways: int = 0
    total_internet_gateways: int = 0
    total_elastic_ips: int = 0
    last_refreshed: Optional[datetime] = None


class TopologyResponse(BaseSchema):
    """Complete topology response for visualization."""

    vpcs: List[TopologyVPC]
    meta: TopologyMeta
